### DISCUSSION TITLE
Some queries get stalled when future=True on the engine

### DISCUSSION THREAD

**Discussion by bigjools** [sqlalchemy/sqlalchemy#6975] [Usage Questions]
I've recently spent a long time converting my medium-sized web application to the v2 API, following the useful guide in the docs. The application runs under Flask and spawns off threaded workers for Spinach in the same process. The DB is Postgres running in REPEATABLE_READ isolation.

For some reason, when future=True when making the engine, after a while things start to gum up because my background jobs start hanging on unfinished queries. I can see from pg_stat_activity that most of them are ROLLBACK. I expect to do a fair few rollbacks anyway because the jobs are highly parallel and can result in DB conflicts due to the isolation level, and I have a retry mechanism that closes the session and starts a new one to replay things. However, it seems as though the rollbacks don't complete and there's quite a few table locks held that end up eventually deadlocking the whole application.

If I leave all my other v2 conversions the same (ie future=True on the sessions, etc), and change the engine back to v1 mode (future=False), everything works just fine.

I am not really sure where to start looking to debug this. Can anyone offer any advice please?

**Comment by zzzeek** [comment]
future mode in the engine does almost nothing except that it will auto begin a transaction before a statement is executed, if no transaction is started.  if all of your engine use is through a Session, this should make no difference, because the session always begins a transaction in any case.  only if you have some direct use of "engine.connect()" with some statements following, will there be a behavioral difference.   running "engine.connect()" and then some statements on that connection will start a transaction, whereas previously it wouldnt.   as long as you .close() that connection immediately, there again shouldnt be much difference, unless you are doing something like running statements that can deadlock on multiple connections at the same time.

if you have direct engine use with engine.connect(), maybe use autocommit isolation level with those, so that there isnt any transaction:

```py
with engine.connect().execution_options(isolation_level='AUTOCOMMIT') as conn:
   # ...
```

**Reply by bigjools** [reply]
Absolutely everything we do is through sessions, there's no direct engine statements at all.

**Reply by CaselIT** [reply]
It would be really useful if you could find a way of reproducing it. I know it may be a bit difficult, since a worker queue is involved.

I'm guessing the application is not open source?

**Reply by zzzeek** [reply]
whats the general session pattern?  session is bound to an engine, and you never call engine.connect() right?

**Reply by zzzeek** [reply]
can you try using future with engine, and make this one patch, then see if issue changes:

```diff
diff --git a/lib/sqlalchemy/engine/base.py b/lib/sqlalchemy/engine/base.py
index a316f904f0..b2947d68df 100644
--- a/lib/sqlalchemy/engine/base.py
+++ b/lib/sqlalchemy/engine/base.py
@@ -1696,8 +1696,8 @@ class Connection(Connectable):
         elif self._trans_context_manager:
             TransactionalContext._trans_ctx_check(self)
 
-        if self._is_future and self._transaction is None:
-            self._autobegin()
+        #if self._is_future and self._transaction is None:
+        #    self._autobegin()
 
         context.pre_exec()
```

that above should be the only place "future" mode does something different that isn't about raising an assertion

**Reply by bigjools** [reply]
@CaselIT Yeah very difficult to make a simple case to reproduce at this time and sadly not open source, no. I'm firefighting my production deployment at the moment and I think we're going to have to turn off all the v2 features. Today I've found out that a sufficient number of worker jobs will cause the same problem even with the v1 engine. It seems specifically related to handling of rollbacks after getting a conflict from the repeatable_read isolation.

**Reply by bigjools** [reply]
@zzzeek No, we're binding the session to the result of engine.connect().

**Reply by bigjools** [reply]
Also I should say that all sessions are binding to the same connection in each thread, and now that I've said that it seems wrong somehow even though that's how I've seen all the examples do it. Should it bind directly to the engine instead?

**Reply by bigjools** [reply]
Ok I have solved all my problems with this bind change. I've no idea where I got the idea that this was the way to do things, but I suspect it started with the changes in requirements around testing fixtures. You might want to add a warning about binding connections in real world use?

Thanks for pointing me in the right direction again.

**Reply by zzzeek** [reply]
> @zzzeek No, we're binding the session to the result of engine.connect().

OK, that's significant, can you confirm that when you bind the session to that connection, connection.in_transactinon() is False?

> Also I should say that all sessions are binding to the same connection in each thread, and now that I've said that it seems wrong somehow even though that's how I've seen all the examples do it. 

can you be more specific?  where is this "connection in each thread" coming from?  do you mean sharing a single connection among many threads?  that woudl be completely wrong.  There should be generally very few examples that bind a Session to a Connection directly, this is an unusual use case, show me these examples.


> Should it bind directly to the engine instead?

absolutely, this is what's documented everywhere.

**Reply by bigjools** [reply]
> where is this "connection in each thread" coming from?

Sorry, I can't English today, was writing that in the middle of a production change. The way things were set up was to match a connection to a single session because each connection was in its own thread, so theoretically it should not have been a problem. Clearly, it was, however :) 

Anyway it's all moot as I have the right bind now. Phew.

**Reply by CaselIT** [reply]
If you could provide a general example on how the connection was used to investigate if that was an usage problem or if there is something else going on. it would be great. 

Was something like this used?

```py
def thread_run():
  conn = engine.connect()
  while True:
    req = wait_for_req()
    with Session(conn) as session:
      handle_req(req, session)

...

Thread(target=thread_run).start()    
```

**Reply by bigjools** [reply]
I'll create an engine and then run

`Session.configure(bind=engine, future=True)`

The _incorrect_ code used `bind=engine.connect()`

Flask is responsible for creating request threads. I also have worker threads created once at startup time. Thereon after it's as simple as using the Session's `__call__` to generate a session where I run `execute` etc., so the returned session object uses the default thread-based factory.

I think because the worker threads run in the application server they were sharing connections with the request threads, so ended up with sessions closing/committing/rolling back each other's connections.

As I said above I think I was initially lead astray by the changed way to handle test fixtures which does need to bind to a connection so that it can do nesting and restarts.

**Reply by CaselIT** [reply]
I assume `Session` is a `sessionmaker` instance, right?

If I'm not mistaken that means that all sessions would use the same underlying connection. 
@zzzeek should we maybe add a warning in the `sessionmaker` when bind / binds are connections?

**Reply by bigjools** [reply]
> I assume `Session` is a `sessionmaker` instance, right?

Correct.

> If I'm not mistaken that means that all sessions would use the same underlying connection.

Yes, I think that is the case too. Desirable for test suites, not so much for production code.

> @zzzeek should we maybe add a warning in the `sessionmaker` when bind / binds are connections?

That would have certainly helped in my case, but unless it can detect it's running as part of a test suite we might also want to be able to turn that off.

**Reply by zzzeek** [reply]
Its a valid pattern if someone knows what they are doing and I don't think it has to warn.  However it's a very strange pattern that's not of general use and in 15 years I've never seen anyone do that
  I've scrolled here to see links to what tutorials would be recommending this as the de facto pattern and I don't see any, but if some tutorial is recommending this it has to be fixed.

**Reply by bigjools** [reply]
As I said above, the thing that originally made me do this was the example at
https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites (apologies for not actually linking it earlier).
It's mostly PEBKAC on my part to be fair, but I did all of this in the context of trying to convert from v1 to v2 style and it was pretty easy to get carried away with the changes required, and in particular the test suite changes clouded my head a lot when it came to dealing with the transaction management changes.

Instead of code warnings, perhaps that section of the document should re-state that in normal operation you don't want to bind to the connection as that's primarily useful only in test fixtures?

**Reply by CaselIT** [reply]
Just to confirm, 1.3 did use the engine in the Session bind?

**Reply by bigjools** [reply]
Yes, that's correct.
